Simplify semaphore move protection

bhalevy opened the pull request:
Referring to #1323 and discussion in #1477.

Avi Kivity <avi@scylladb.com> asserted that:
> We have a choice:
> declare that most things can be movable and live with the complication and slowdown
> declare that control structures (like semaphores) have stable addresses (as they typically do anyway)
> There's really no reason to move a sempahore.

Gleb Natapov <gleb@scylladb.com> proposed:
> add an assertion in debug mode if a semaphore is moved after the first use. It can be a simple bool that is set to true on the first wait().

Since we still need to be able to move the semaphore
after it created this patch adds `bool _used` and implements
the move ctor, asserting that the semaphore is moved before used.

Close #1323

gleb-cloudius commented on include/seastar/core/semaphore.hh:
```diff
@@ -154,6 +154,7 @@ public:
 private:
     ssize_t _count;
     std::exception_ptr _ex;
+    bool _used;
```
Shouldn't we do in debug mode only like the reverted patch did?

bhalevy replied:
makes sense

bhalevy commented:
In v2 ([359a177](https://github.com/scylladb/seastar/commit/359a177dbf1c865d7209e449975ddfbc30a2f375)):
- restricted change to SEASTAR_DEBUG

gleb-cloudius commented on include/seastar/core/semaphore.hh:
```diff
@@ -276,6 +319,7 @@ public:
     ///         \ref semaphore_aborted exception.  If the semaphore was
     ///         \ref broken(), may contain a \ref broken_semaphore exception.
     future<> wait(abort_source& as, size_t nr = 1) noexcept {
+        use();
```
What about `try_wait()` should we add it there as well?

bhalevy replied:
It never creates a waiter so it's not strictly necessary.

bhalevy replied:
Hmm, strike that out.  It's called from `try_get_units` that creates semaphore_units that depends on the semaphore, so it is needed after all.

bhalevy replied:
Same goes for consume, being called from `consume_units`...

bhalevy commented:
in v3 ([e01afe5](https://github.com/scylladb/seastar/commit/e01afe588266d358fca9de212a66f839853fec4e)):
- added `use()` calls in `try_wait` and `consume`.

bhalevy commented:
CI failure in spawn test looks unrelated.
I'll rerun it.

bhalevy commented:
> CI failure in spawn test looks unrelated. I'll rerun it.

Well, I don't have permissions to re-run it.

@scylladb/seastar-maint can you please rerun the ci job?

bhalevy commented:
same failure in ci seen on another PR too, see https://app.circleci.com/pipelines/github/tchaikov/seastar/468/workflows/c6568558-6508-46f8-9f87-433f1e8231ed/jobs/5604

nyh commented:
> > CI failure in spawn test looks unrelated. I'll rerun it.
> 
> Well, I don't have permissions to re-run it.
> 
> @scylladb/seastar-maint can you please rerun the ci job?

I don't even have a circleci account... Who is responsible for circleci?

benipeled commented:
> > > CI failure in spawn test looks unrelated. I'll rerun it.
> > 
> > 
> > Well, I don't have permissions to re-run it.
> > @scylladb/seastar-maint can you please rerun the ci job?
> 
> I don't even have a circleci account... Who is responsible for circleci?

@avikivity isn't you enable the _"Build forked pull requests"_ option for the seastar project?
If not, please enable it to avoid the need of circleci account (for contributors)

tchaikov commented:
> same failure in ci seen on another PR too, see https://app.circleci.com/pipelines/github/tchaikov/seastar/468/workflows/c6568558-6508-46f8-9f87-433f1e8231ed/jobs/5604

sorry for the flaky test. i am disabling it in #1498 .

avikivity commented:
We lost the test, no?

avikivity commented on include/seastar/core/semaphore.hh:
```diff
@@ -214,6 +232,30 @@ public:
+        , _used(o._used)
+    {
+        assert(!_used && "semaphore cannot be moved after it has been used");
+    }
```
You can have a class used_flag { bool used } to avoid writing the complex constructors/assignops and focus only on the used flag. It will reduce complexity and duplication.

bhalevy replied:
done in v4

bhalevy commented:
In v4 ([eb8dfc7](https://github.com/scylladb/seastar/commit/eb8dfc740d436b00531782ea63bc9935b485ed9b)):
- added `struct used_flag` that narrows down to scope for manipulating and checking the _used state to this class only.

tchaikov commented on include/seastar/core/semaphore.hh:
```diff
@@ -210,14 +189,41 @@ private:
+#ifdef SEASTAR_SEMAPHORE_DEBUG
+    struct used_flag {
+        // set to true from the wait path
+        // prevents the semaphore to be moved or move-reassigned when _used
```
nit

```suggestion
        // prevents the semaphore from being moved or move-reassigned when _used
```

bhalevy replied:
updated, thanks

bhalevy replied:
in v5

avikivity commented on include/seastar/core/semaphore.hh:
```diff
@@ -185,6 +189,35 @@ private:
+    };
+#endif
+
+    used_flag _used;
```
Add [[no_unique_address]] to avoid increasing the semaphore size in non-debug builds.

bhalevy replied:
in v5

avikivity commented:
> We lost the test, no?

^

Suggest merging the two patches so we can keep the test unchanged.

bhalevy commented:
In v5 ([de9fc12](https://github.com/scylladb/seastar/commit/de9fc121425ec14d9bbbdbab00c5b1a43d9e3771)):
- rebased + retested
- fixed comment per https://github.com/scylladb/seastar/pull/1480#discussion_r1103821486- added `[[no_unique_address]]` per https://github.com/scylladb/seastar/pull/1480#discussion_r1103848005
- patches merged, while keeping the unit test per https://github.com/scylladb/seastar/pull/1480#issuecomment-1427093290

kbr-scylla commented:
The original patch now got reverted
```
commit 55681bfde4c8bbc689bf93aa1b11b0eef97bb72f (origin/master, origin/HEAD)
Author: Avi Kivity <avi@scylladb.com>
Date:   Mon Feb 13 16:32:55 2023 +0200

    Revert "semaphore: assert no outstanding units when moved"
    
    This reverts commit 3061b2f53a9a89596786763db03e3ae44398d33c. It has
    been shown to cause numerous problems in scylladb/scylladb, so it
    must be broken in some way.
```

so this will require a rebase.
